planner/function_dependency/fast_int_set.go

```diff
@@ -0,0 +1,376 @@
+package function_dependency
+
+import (
+	"bytes"
+	"fmt"
+	"golang.org/x/tools/container/intsets"
+	"math/bits"
+)
+
+const smallCutOff = 64
+
+// FastIntSet is wrapper of sparse with an optimization that number (0 ~ 63) can be cached for quick access.
+// From the benchmark in fd_graph_test.go, we choose to use sparse to accelerate int set. And when the set
+// size is quite small we can just skip the block allocation in the sparse chain list.
+type FastIntSet struct {
+	// an uint64 used like quick-small bitmap of 0~63.
+	small uint64
+	// when some is bigger than 64, then all that previous inserted will be dumped into sparse.
+	large *intsets.Sparse
+}
+
+// NewFastIntSet is used to make the instance of FastIntSet with initial values.
+func NewFastIntSet(values ...int) FastIntSet {
+	var res FastIntSet
+	for _, v := range values {
+		res.Insert(v)
+	}
+	return res
+}
+
+// Len return the size of the int set.
+func (s FastIntSet) Len() int {
+	if s.large == nil {
+		return bits.OnesCount64(s.small)
+	}
+	return s.large.Len()
+}
+
+// Insert is used to insert a value into int-set.
+func (s *FastIntSet) Insert(i int) {
+	isSmall := i >= 0 && i < smallCutOff
+	if isSmall {
+		s.small |= 1 << uint64(i)
+	}
+	if !isSmall && s.large == nil {
+		// first encounter a larger/smaller number, dump all that in `small` into `large`.
+		s.large = s.toLarge()
+	}
+	if s.large != nil {
```

Comment:
When there's the large part, we would insert each small value twice both in the small part and the large part?

Reply:
yes, twice exactly